File: reverse-in-parens/c/commented.c
#include <string.h>
#include <stdlib.h>

char * reverse_in_parens(const char * text){
    // (dangerously) copy text into output
    char * out = strcpy(malloc(strlen(text) + 1), text);
    // iterate through string.
    for(int i=0, end_paren=0; out[i] != '\0'; i++){
        // skip all non opening parens
        if(out[i] == '('){
            // Iterate forward to find end paren. Track number of
            // open parens as open.
            for(int j=i+1, open = 1; out[j] != '\0'; j++){
                // mark new open parens
                if(out[j] == '(') open++;
                // reduce parens if they close, set end_paren
                // if all closed and break;
                else if(out[j] == ')' && (--open) == 0 && (end_paren = j)) break;
            }
            // iterate forward to actually reduce text, from 1 char
            // after opening paren to 1 char after closing.
            for(int j=i+1, r=end_paren-1; j<end_paren-(end_paren-i-1)/2; j++, r--){
                char tmp = out[j];
                // Copy out[r] to out[j], and reverse them if they are parens
                out[j] = (out[r]=='(')?')':(out[r]==')')?'(':out[r];
                // do the same for out[j] to out[r], but skip if they
                // point to the same char.
                if(j != r) out[r] = (tmp=='(')?')':(tmp==')')?'(':tmp;
            }
        }
    }
    return out;
}

File: reverse-in-parens/js/commented.js
function reverseInParens(text, startWithReverse = false){
    // Helper function which reverses the string, but
    // swaps forwards and backwards parentheses.
    const reverseParens = t =>{
        t.reverse();
        for(let i=0; i<t.length; i++){
            if(t[i] === '(') t[i] = ')';
            else if(t[i] === ')') t[i] = '(';
        }
        return t;
    }

    // Helper function to find the end parenthese
    const findEndParen = (t, i=0) => {
        let open = 1;
        for(; i<t.length; i++){
            if(t[i] === '(') open++;
            else if(t[i] === ')'){
                open--;
                if(open === 0) return i;
            }
        }
        return -1;
    }

    // Recursive calls will pass an array, but
    // the initial call may be a string, so we
    // array-ify it here.
    if(typeof text === 'string') text = text.split('');
    // indicates if we should reverse the space in the string outside parens
    if(startWithReverse) text = reverseParens(text, true);
    let endParen = 0;
    // Iterate through each instance of opening parens
    for(let startParen = text.indexOf('('); startParen !== -1; startParen = text.indexOf('(', endParen)){
        // get index of the end paren with relation to startParen
        endParen = findEndParen(text, startParen + 1);
        // Run this function on only the text inside the parens. The true parameter
        // indicates that all text should be immediately reversed before further parsing.
        let rev = reverseInParens(text.splice(startParen + 1, endParen - startParen - 1), true);
        // add output from our recursive call to the output
        text.splice(startParen+1, 0, ...rev);
    }
    return (typeof text === 'object') ? text.join('') : text;
}